### ISSUE TITLE
Guide on Using With GitHub Actions

### ISSUE THREAD

**Issue by alexfinnarn** [salesforce/codeceptjs-bdd#55]
Hello, I've been trying to get a sensible working solution for end-to-end testing within Salesforce and have hit many roadblocks. 

I had a version of code where I could reliably use the Travis CI free plan and Cypress to test the older VisualForce/Lightning UIs, but then came LWC and the terrible Shadow DOM. Also, then Travis CI stopped providing an open-source tier so I had to go back to GitHub Actions...and then the Salesforce Trailhead/docs to get Travis CI and the token, etc. working were removed.

So, now I am glad to find this solution which I can use locally just fine. However, I can't seem to get the JWT auth flow working on GitHub Actions. Since this project looks to be the defacto/future way to do E2E tests with LWC and Salesforce, I'm hoping a guide on doing this in GitHub, also Salesforce recommended, can be added to the docs.

I think Salesforce has a deal with CircleCI or something...but I can be of assistance in creating such a guide for GitHub Actions. I have a sample repo and current "To access this page, you have to log in to Salesforce." errors in logs.

So my three questions:
- Is this the defacto way to do E2E testing in Salesforce going forward? Or at least Salesforce will support testing LWC (non-unit tests) through this project?
- Is there a recommended way to use this project on CI platforms, specifically GitHub Actions/Workflows?
- Is this the best way to "log in" during an E2E test? https://github.com/cypress-io/cypress/issues/2367#issuecomment-584058927

Sorry to ask these questions in a generic issue, and please point me to a better forum, if the discussion should not be on within GitHub issues.

**Comment by gkushang** [comment]
@alexfinnarn thanks for reaching out. 

yes, it'd be great to have guide on using Github Actions!! 

**Regarding this framework:**

We built this solution specifically to deal with LWC (Shadow Elements) for Salesforce E2E tests running on WebDriver. Webdriver does have any official solution yet to locate Shadow DOM at ease or not even a standard yet with W3C or from browser vendors - so locating Shadow DOMs is pain, but this framework solves this issue.

We have built and extended WebDriver API to locate Shadow DOM at ease. E.g.

```
I.click('.any-shadow-elements-in-dom'); // no dom hierarchy is required with this framework
```
For more information on How this framework deal with Shadow DOMs in WebDriver, pls take a look at this [Salesforce Engineering Blog/working-with-shadow-doms-webdrivers](https://engineering.salesforce.com/automation-made-easy-for-web-components-shadow-roots-b23016779ce8)

We use this framework primarily to automate Two areas of UI automation, 

- **Low level UI Integration tests (with Playwrights)**
   - running Headless on [Playwright browsers](https://github.com/microsoft/playwright) on both Desktop and Mobile/Tablet layouts. Playwright internally supports the Shadow DOMs unlike Webdrivers.
- **E2E testing (with WebDrivers)**
   - This framework extends the WebDriver API to deal with Shadow DOM as mentioned above.
      - Take a look at the Example of [WebDriver Tests for WebComponents / Shadow DOMS](https://github.com/salesforce/codeceptjs-bdd/blob/develop/examples/webcomponents-playwright-webdriver-example/README.md)
   - This framework is already integrated with Sauce Labs and Selenoid Grid. To run on either platform, the existing commands are,
       - `yarn ui:test --profile sauce:config:'macOS 10.15':safari:13.`
       - `yarn ui:test --profile selenoid:chrome:80`

**Advantage:** Same tests can be run on both Playwright and Webdriver. So, your UI Integration tests can be extended to write more E2E tests without any changes

**Regarding your questions**
> Is this the defacto way to do E2E testing in Salesforce going forward? Or at least Salesforce will support testing LWC (non-unit tests) through this project?

This framework primarily supports locating LWC Shadow DOM's for UI Testing at ease. If you have use-cases for UI Integration and E2E tests, this framework could be one of the good choice for you

> Is there a recommended way to use this project on CI platforms, specifically GitHub Actions/Workflows?

This framework does not limit the use of CI platform, so yes! this framework is currently integrated with [Circle CI](https://github.com/salesforce/codeceptjs-bdd/blob/develop/.circleci/config.yml) where we run our CI tests. 


> Is this the best way to "log in" during an E2E test? cypress-io/cypress#2367 (comment)

I have not tried this use case, but based on the solution you've mentioned, I believe as long as SFDX CLI is installed the `shelljs` can execute it and give you the instanceUrl which eventually can be used to navigate through `I.amOnPage(instanceUrl)`. I'd love to hear how it goes if you'r planning to try it out.

Pls feel free to reach out if you have any further questions!

**Comment by gkushang** [comment]
Hope the above answers your question. Feel free to reach out to us for any more questions or concerns.

**Comment by alexfinnarn** [comment]
Thanks for the reply! I've been busy working on other projects, but I will get back to this soon and try out the `shelljs` command. 

This issue was supposed to be a placeholder for adding to the docs. I know some projects have a separate way of doing that than in the code repo and related issues. I can reopen when I have some kind of solution to provide some docs for, but I wanted to add it just in case someone else comes in here looking for the same thing.

I'll see what I can find out...

**Comment by alexfinnarn** [comment]
After adding `shelljs`, I was able to log in with the following snippet. I ran the rest of the code within that command callback vs. async/await or some more fancy coding.

```js
  shell.exec(`sfdx force:org:display -u ${process.env.SF_USER} --json`, function(code, stdout, stderr) {
    const out = JSON.parse(stdout);
    I.amOnPage(`${instanceUrl}/secur/frontdoor.jsp?sid=${out.result.accessToken}`);
    // ...other code.
  });
```

My work, I think, is pulling me away from completing the task I was working on, but I wanted to add that snippet. Now, I am back on the `sfdx force:auth:jwt:grant` step which I had working on Travis CI and some place else...but forgot to write the steps exactly down :(

Maybe this helps somebody, maybe not...but good luck with this and I'll check back from time to time.
